test/sparsedir/sparse.jl

```diff
@@ -288,6 +288,60 @@ end
 cA = sprandn(5,5,0.2) + im*sprandn(5,5,0.2)
 @test full(conj(cA)) == conj(full(cA))
 
+# Test SparseMatrixCSC [c]transpose[!] and permute[!] methods
+let smalldim = 5, largedim = 10, nzprob = 0.4
+    (m, n) = (smalldim, smalldim)
+    A = sprand(m, n, nzprob)
+    X = similar(A)
+    C = transpose(A)
+    p = randperm(m)
+    q = randperm(n)
+    # Test common error checking of [c]transpose! methods (ftranspose!)
+    @test_throws DimensionMismatch transpose!(A[:, 1:(smalldim - 1)], A)
+    @test_throws DimensionMismatch transpose!(A[1:(smalldim - 1), 1], A)
+    @test_throws ArgumentError transpose!((B = similar(A); resize!(B.rowval, nnz(A) - 1); B), A)
+    @test_throws ArgumentError transpose!((B = similar(A); resize!(B.nzval, nnz(A) - 1); B), A)
+    # Test common error checking of permute[!] methods / source-perm compat
+    @test_throws DimensionMismatch Base.SparseArrays.permute(A, p[1:(end - 1)], q)
```

Comment:
Now it can definitely go. I think the "getindex can't know it's a permutation until we have a type for that" has informally sold over at least Keno and probably others around the Julia Computing office today.